tests/Tests/QuickCheckUtils.hs

```diff
@@ -240,33 +246,36 @@ instance Arbitrary IO.BufferMode where
 --   sometimes contain line endings.)
 -- * Newline translation mode.
 -- * Buffering.
-write_read :: (NFData a, Eq a, Show a)
-           => ([b] -> a)
-           -> ((Char -> Bool) -> a -> b)
-           -> (IO.Handle -> a -> IO ())
-           -> (IO.Handle -> IO a)
-           -> IO.NewlineMode
-           -> IO.BufferMode
-           -> [a]
-           -> Property
-write_read _ _ _ _ (IO.NewlineMode IO.LF IO.CRLF) _ _ = discard
-write_read unline filt writer reader nl buf ts = ioProperty $
-    (===t) <$> act
+write_read :: forall a b c.
+  (Eq a, Show a, Show c, Arbitrary c)
+  => ([b] -> a)
+  -> ((Char -> Bool) -> b -> b)
+  -> (IO.Handle -> a -> IO ())
+  -> (IO.Handle -> IO a)
+  -> (c -> [b])
+  -> [TestTree]
+write_read unline filt writer reader modData
+  = encodings <&> \enc@TextEncoding {textEncodingName} -> testGroup textEncodingName
+    [ testProperty "NoBuffering" $ propTest enc (\() -> IO.NoBuffering) (const False)
+    , testProperty "LineBuffering" $ propTest enc (\() -> IO.LineBuffering) (const False)
+    , testProperty "BlockBuffering" $ propTest enc IO.BlockBuffering (maybe False (< 4))
+    ]
   where
-    t = unline . map (filt (not . (`elem` "\r\n"))) $ ts
-
-    act = withTempFile $ \path h -> do
-            IO.hSetEncoding h IO.utf8
-            IO.hSetNewlineMode h nl
-            IO.hSetBuffering h buf
-            () <- writer h t
-            IO.hClose h
-            bracket (IO.openFile path IO.ReadMode) IO.hClose $ \h' -> do
-              IO.hSetEncoding h' IO.utf8
-              IO.hSetNewlineMode h' nl
-              IO.hSetBuffering h' buf
-              r <- reader h'
-              r `deepseq` return r
+  propTest :: TextEncoding -> (modeArg -> IO.BufferMode) -> (modeArg -> Bool) -> modeArg -> IO.NewlineMode -> c -> Property
+  propTest _   _               _           _      (IO.NewlineMode IO.LF IO.CRLF) _ = discard
+  propTest _   _               modeArgPred bufArg _                              _ | modeArgPred bufArg = discard
+  propTest enc mkBufferingMode _           bufArg nl                             d = ioProperty $ withTempFile $ \_ h -> do
+    let ts = modData d
+        t = unline . map (filt (not . (`elem` "\r\n"))) $ ts
+    IO.hSetEncoding h enc
+    IO.hSetNewlineMode h nl
+    IO.hSetBuffering h $ mkBufferingMode bufArg
+    () <- writer h t
+    IO.hSeek h IO.AbsoluteSeek 0
+    r <- reader h
+    let isEq = r == t
+    deepseq isEq $ pure $ counterexample (show r ++ bool " /= " " == " isEq ++ show t) isEq
```

Comment:
```suggestion
    pure (r === t)
```

`isEq` is a boolean which is about to be forced by `counterexample`, so `seq`/`deepseq` does nothing. And `===` in QuickCheck abbreviates `counterexample` of `==`.

Reply:
I was having problems with lazy IO of LazyText and the handle closing before equality was tested.

I think `pure (r === t)` just passes the Property back but does not evaluate `r == t` until the handle is already closed.

Reply:
I see. Since the real culprit is lazy IO, could this be fixed at the point where `write_read` is applied to lazy IO functions (wrap them to make them strict) instead of inside `write_read`?

Reply:
I think moving strictness responsibly to the caller makes the interface harder to use since it is not captured in the type signature. Also, I am planning on adding Builder to this and cannot deep sequence it but I can check for equality.

Reply:
ok.
it should be seq though.